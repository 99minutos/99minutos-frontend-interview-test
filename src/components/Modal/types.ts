export type ModalProps = {
  handleCloseModal: () => void
}