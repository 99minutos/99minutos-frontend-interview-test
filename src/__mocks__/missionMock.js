export default {
  mission_name: '',
  details: null,
  links: {
    video_link: '',
    flickr_images: [
      ''
    ],
  }
};