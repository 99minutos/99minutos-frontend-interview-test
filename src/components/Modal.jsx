import React, { useState } from 'react';
import ReactDOM from 'react-dom';
import '../css/components/Modal.css';
import astronaut from '../images/astronaut.jpeg'

const Modal = ({ isOpen, onClose, missionSelected }) => {

  if(!isOpen) {
    return null
  }
  return ReactDOM.createPortal(
    <div className="Modal">
      <div className="Modal--container">
        <button onClick={onClose} className="Modal--close-button">
            X
        </button>
        <img src={missionSelected.links.flickr_images[0] ?? astronaut} alt=""/>
        <div className="Modal--content">
          <h2>{missionSelected.mission_name}</h2>
          <p>{missionSelected.details || 'Details omitted'}</p>
          <a target="_blank" href={missionSelected.links.video_link}>See more</a>
        </div>
      </div>
    </div>,
    document.getElementById('modal')
  );
};

export default Modal;