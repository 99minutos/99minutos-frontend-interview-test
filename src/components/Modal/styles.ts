import styled from 'styled-components';

import { colors } from '../../assets/styles/config';
import { Cross } from '../../assets/images';

export const ModalStyled = styled.div`
  position: fixed;
  top: 0;
  z-index: 2;
  width: 100%;
  height: 100%;
  background-color: gray;
`;

export const CloseButton = styled.button`
  position: absolute;
  z-index: 3;
  top: 20px;
  right: 26px;
  border: none;
  outline: none;
  width: 70px;
  height: 70px;
  padding: 12px;
  border-radius: 50%;
  box-shadow: 0 2px 5px ${colors.shadowDark};
  background-color: ${colors.midNightBlue};

  &:active {
    opacity: 0.8;
    transform: scale(0.96);
  }
`;

export const CloseIcon = styled(Cross)`
  max-width: 100%;
  max-height: 100%;
  fill: ${colors.white};
`;