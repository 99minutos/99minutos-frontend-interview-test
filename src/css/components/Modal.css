.Modal {
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  background-color: rgba(0, 0, 0, 0.5);
  display: flex;
  flex-direction: column;
  align-items: center;
  color: var(--color-black-primary);
}

.Modal--container {
  position: relative;
  top: 2rem;
  background-color: #ffffff;
  width: 500px;
}

.Modal--container img {
  width: 100%;
  height: 300px;
}

.Modal--content {
  padding: 1em;
  text-align: center;
}

.Modal--content h2 {
  margin-bottom: 10px;
}

.Modal--content p {
  text-align: justify;
  margin-bottom: 30px;
}

.Modal--content a {
  padding: 5px;
  color: #ffffff;
  background-color: blueviolet;
  text-decoration: none;
}

.Modal--close-button {
  position: absolute;
  top: 0;
  right: 0;
  border: 0;
  background-color: #F5F5F5;
  padding: 0.5rem 1rem;
  cursor: pointer;
  outline-style: none;
}