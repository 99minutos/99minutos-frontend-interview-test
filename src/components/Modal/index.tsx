import React, { FC, useEffect, useState } from 'react';
import { createPortal } from 'react-dom';

import { ModalProps } from './types';
import { CloseButton, CloseIcon, ModalStyled } from './styles';

const Modal: FC<ModalProps> = ({ children, handleCloseModal }) => {

  const [domElement, setDomElement] = useState<any>(() => {
    return document.getElementById('modal');
  });

  useEffect(() => {
    return () => setDomElement(() => null);
  },
  [])

  return createPortal(
    <ModalStyled>
      <CloseButton onClick={handleCloseModal}>
        <CloseIcon />
      </CloseButton>
      {children}
    </ModalStyled>,
    domElement
  )
}

export default Modal;