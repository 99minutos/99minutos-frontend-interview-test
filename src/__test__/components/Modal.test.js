import React from 'react';
import { mount } from 'enzyme';
import { create } from 'react-test-renderer';
import Modal from '../../components/Modal';
import missionMock from '../../__mocks__/missionMock';
import { expect, jest } from '@jest/globals';
import ReactDOM from 'react-dom';

describe('<Modal />', () => {
  let component;

  // add a div with #modal-root id to the global body
  const modalRoot = global.document.createElement('div');
  modalRoot.setAttribute('id', 'modal');
  const body = global.document.querySelector('body');
  body.appendChild(modalRoot);
  test('Component Modal render', () => {
    component = mount(<Modal missionSelected={missionMock} isOpen />);
    expect(component.length).toEqual(1);
  });

  test('It not received the mission details props', () => {
    component = mount(<Modal missionSelected={missionMock} isOpen />);
    expect(component.find('p').text()).toBe('Details omitted');
  });

  test('Should trigger onClose when clicked', () => {
    const onClose = jest.fn();
    component = mount(<Modal missionSelected={missionMock} isOpen onClose={onClose} />);

    component.find('.Modal--close-button').simulate('click');
    expect(onClose.mock.calls).toHaveLength(1)
  })

  afterEach(() => {
    component.unmount();
  });
});

describe('Modal Snapshot', () => {
  beforeAll(() => {
    ReactDOM.createPortal = jest.fn((element, node) => {
      return element
    });
  });

  afterEach(() => {
    ReactDOM.createPortal.mockClear()
  });
  test('Check UI of component Modal', () => {
    const modal = create(<Modal missionSelected={missionMock} isOpen />);
    expect(modal.toJSON()).toMatchSnapshot();
  });
});