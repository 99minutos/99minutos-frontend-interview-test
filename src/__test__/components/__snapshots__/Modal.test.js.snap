// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Snapshot Check UI of component Modal 1`] = `
<div
  className="Modal"
>
  <div
    className="Modal--container"
  >
    <button
      className="Modal--close-button"
    >
      X
    </button>
    <img
      alt=""
      src=""
    />
    <div
      className="Modal--content"
    >
      <h2>
        
      </h2>
      <p>
        Details omitted
      </p>
      <a
        href=""
        target="_blank"
      >
        See more
      </a>
    </div>
  </div>
</div>
`;
